dashboard/src/app/index.module.ts

```diff
@@ -36,12 +36,81 @@ import IdeIFrameSvc from './ide/ide-iframe/ide-iframe.service';
-import {WorkspaceDetailsService} from './workspaces/workspace-details/workspace-details.service';
+import {OrganizationsConfig} from './organizations/organizations-config';
+import {TeamsConfig} from './teams/teams-config';
 
 // init module
 const initModule = angular.module('userDashboard', ['ngAnimate', 'ngCookies', 'ngTouch', 'ngSanitize', 'ngResource', 'ngRoute',
   'angular-websocket', 'ui.bootstrap', 'ui.codemirror', 'ngMaterial', 'ngMessages', 'angularMoment', 'angular.filter',
-  'ngDropdowns', 'ngLodash', 'angularCharts', 'uuid4', 'angularFileUpload']);
+  'ngDropdowns', 'ngLodash', 'angularCharts', 'uuid4', 'angularFileUpload', 'ui.gravatar']);
+
+declare const Keycloak: Function;
+function buildKeycloakConfig(keycloakSettings: any) {
+  return {
+    url: keycloakSettings['che.keycloak.auth_server_url'],
+    realm: keycloakSettings['che.keycloak.realm'],
+    clientId: keycloakSettings['che.keycloak.client_id']
+  };
+}
+interface IResolveFn<T> {
+  (value: T | PromiseLike<T>): Promise<T>;
+}
+interface IRejectFn<T> {
+  (reason: any): Promise<T>;
+}
+function keycloakLoad(keycloakSettings: any) {
+  return new Promise((resolve: IResolveFn<any>, reject: IRejectFn<any>) => {
+    const script = document.createElement('script');
+    script.async = true;
+    script.src = keycloakSettings['che.keycloak.auth_server_url'] + '/js/keycloak.js';
+    script.addEventListener('load', resolve);
+    script.addEventListener('error', () => reject('Error loading script.'));
+    script.addEventListener('abort', () => reject('Script loading aborted.'));
+    document.head.appendChild(script);
+  });
+}
+function keycloakInit(keycloakConfig: any) {
+  return new Promise((resolve: IResolveFn<any>, reject: IRejectFn<any>) => {
+    const keycloak = Keycloak(keycloakConfig);
+    keycloak.init({
+      onLoad: 'login-required', checkLoginIframe: false
+    }).success(() => {
+      resolve(keycloak);
+    }).error((error: any) => {
+      reject(error);
+    });
+  });
+}
+
+const keycloakAuth = {
+  isPresent: false,
+  keycloak: null,
+  config: null
+};
+initModule.constant('keycloakAuth', keycloakAuth);
+
+const promise = new Promise((resolve: IResolveFn<any>, reject: IRejectFn<any>) => {
+  angular.element.get('/api/keycloak/settings').then(resolve, reject);
+});
+promise.then((keycloakSettings: any) => {
+  keycloakAuth.config = buildKeycloakConfig(keycloakSettings);
+
+  // load Keycloak
+  return keycloakLoad(keycloakSettings).then(() => {
+    // init Keycloak
+    return keycloakInit(keycloakAuth.config);
+  }).then((keycloak: any) => {
+    keycloakAuth.isPresent = true;
+    keycloakAuth.keycloak = keycloak;
+    /* tslint:disable */
+    window['_keycloak'] = keycloak;
+    /* tslint:enable */
+  });
+}).catch((error: any) => {
+  console.error('Keycloak initialization failed with error: ', error);
+}).then(() => {
+  angular.bootstrap(document.body, ['userDashboard'], {strictDi: true}); // manually bootstrap Angular
```

Comment:
can this logic be moved instead to the routeProvider initialization
like
https://github.com/eclipse/che/blob/fb48c3d5d6942bd9fcc02f23f511c5b18743b35a/dashboard/src/app/index.module.ts#L118-L135

where we get user preferences before the route changes

it would be more compliant with the existing solution used

Reply:
@davidfestal @akurinnoy can you comment?

Reply:
@benoitf AFAIK no
because application bootstrapping starts routeProvider initialization

Reply:
but routeProvider initialization can be delayed. This is what is done in the snippet

Reply:
> resolve - {Object.<string, Function>=} - An optional map of dependencies which should be injected into the controller. If any of these dependencies are promises, the router will wait for them all to be resolved or one to be rejected before the controller is instantiated. If all the promises are resolved successfully, the values of the resolved promises are injected and $routeChangeSuccess event is fired. If any of the promises are rejected the $routeChangeError event is fired. For easier access to the resolved dependencies from the template, the resolve map will be available on the scope of the route, under $resolve (by default) or a custom name specified by the resolveAs property (see below). This can be particularly useful, when working with components as route templates.